libtiledbsoma/src/reindexer/reindexer.cc

```diff
@@ -0,0 +1,202 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines the a pybind11 api into SOMA C++ library.
+ */
+
+#include "reindexer.h"
+#include <unistd.h>
+#include "khash.h"
+
+uint64_t current_time() {
+    uint64_t microseconds_since_epoch =
+        std::chrono::duration_cast<std::chrono::microseconds>(
+            std::chrono::system_clock::now().time_since_epoch())
+            .count();
+    return microseconds_since_epoch;
+}
+
+static std::mutex log_mutex;
+
+static bool no_logging_s = true;
+
+static void log(std::string message, int ln, std::string severity = "INFO") {
+    if (!no_logging_s) {
+        std::lock_guard l(log_mutex);
+        std::cout << current_time() << " LN " << std::to_string(ln) << " LOG("
+                  << severity << "): " << message << std::endl;
+    }
+}
+
+void IntIndexer::map_locations(const int64_t* keys, int size, int threads) {
+    log("Start of Map locations [" + std::to_string(long((void*)keys)) +
+            "] of size " + std::to_string(size) + " and " +
+            std::to_string(threads) + " threads ",
+        __LINE__);
+    hash_ = kh_init(m64);
+    kh_resize(m64, hash_, size * 1.25);
+    log("Thread pool started and hash table created ", __LINE__);
+    int ret;
+    khint64_t k;
+    int64_t counter = 0;
+    // Hash map construction
+    for (int i = 0; i < size; i++) {
+        k = kh_put(
+            m64,
+            hash_,
+            keys[i],
+            &ret);  // hash_wrapper<IntType>::put(hash_, key, &ret);
+        assert(k != kh_end(hash_));
+        kh_val(hash_, k) = counter;
+        counter++;
+    }
+    size_ = size;
+    thread_count_ = threads;
+    tiledb_thread_pool_ = std::make_unique<tiledbsoma::ThreadPool>(
+        thread_count_);
+
+    log("End of Map locations of size " + std::to_string(size) + " and " +
+            std::to_string(threads) + " threads ",
+        __LINE__);
+}
+
+int IntIndexer::size() {
+    return size_;
+}
+
+std::vector<int64_t> IntIndexer::serial_lookup(
+    const std::vector<int64_t>& keys) {
+    std::vector<int64_t> results;
+    results.resize(keys.size());
+    for (std::vector<int64_t>::size_type i = 0; i < keys.size(); i++) {
+        auto k = kh_get(
+            m64,
+            hash_,
+            keys[i]);  // hash_wrapper<IntType>::get(hash_, keys[i]);
+        if (k == kh_end(hash_)) {
+            throw std::runtime_error(
+                "Key " + std::to_string(keys[i]) +
+                " was not found in the hash table");
+        }
+        results[i] = kh_val(hash_, k);
+    }
+
+    return results;
+}
+
+void IntIndexer::lookup(int64_t* keys, int64_t* results, int size) {
+    log("Lookup with threadcount " + std::to_string(thread_count_) +
+            " on data size of " + std::to_string(size),
+        __LINE__);
+    if (thread_count_ == 1) {
```

Comment:
Can we avoid the optimization of the single thread. Keeping the code with a single code path improve readability, maintainability and test-ability.

Reply:
The only reason for having this path is for potential testing.